src/main/java/co/cask/hydrator/plugin/batch/aggreagtor/aggregator/Sampling.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.aggreagtor.aggregator;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchAggregator;
+import co.cask.cdap.etl.api.batch.BatchAggregatorContext;
+import org.apache.commons.collections.IteratorUtils;
+import org.apache.commons.collections.buffer.PriorityBuffer;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Random;
+import javax.annotation.Nullable;
+
+/**
+ * Sampling plugin to sample random data from large dataset flowing through the plugin.
+ */
+@Plugin(type = BatchAggregator.PLUGIN_TYPE)
+@Name("Sampling")
+@Description("Sampling a large dataset flowing through this plugin to pull random records.")
+public class Sampling extends BatchAggregator<String, StructuredRecord, StructuredRecord> {
+    private enum TYPE {
+        SYSTEMATIC, RESERVOIR
+    }
+
+    private SamplingConfig config;
+
+    public Sampling(SamplingConfig config) {
+        this.config = config;
+    }
+
+    @Override
+    public void prepareRun(BatchAggregatorContext context) throws Exception {
+        context.setNumPartitions(1);
+    }
+
+    @Override
+    public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+        config.validate();
+        pipelineConfigurer.getStageConfigurer()
+                .setOutputSchema(pipelineConfigurer.getStageConfigurer().getInputSchema());
+    }
+
+    @Override
+    public void groupBy(StructuredRecord record, Emitter<String> emitter) throws Exception {
```

Comment:
To group the data for sampling algorithms defined in aggregate(), this function is needed. Please suggest, if i am missing anything.